src/main/groovy/net/vivin/gradle/versioning/VersionUtils.groovy

```diff
@@ -140,7 +144,37 @@ class VersionUtils {
             refresh()
         }
 
-        return versions?.find()
+        if(!versionByTag) {
+            return null
+        } else {
+            def revWalk = new RevWalk(repository)
+            def candidateTags = []
+            def toInvestigateForTags = []
+
+            try {
+                toInvestigateForTags.add(revWalk.parseCommit(repository.resolve(Constants.HEAD)))
+
+                for(RevCommit investigatee = toInvestigateForTags.pop();
```

Comment:
That's a fair point - wish I had seen this comment earlier before I made the changes.

I understand the argument about broadening the scope, but in this particular case I was able to still do it without any broadening, since my stopping condition is when the frontier set (`toInvestigateForTags` in this case) is empty. This way the pointer to the node (`investigatee`) is still scoped to the body of the loop. 

Another argument, and this may be a personal opinion, is that I find `while` more readable in certain cases. So for example if I had `while(!toInvestigateForTags.empty())`, in my head I am able to directly translate it to "while we still have tags to investigate". So the stopping condition (which I think is more relevant in this case) is right up front. 

I find for-loops more useful in bounded situations where you know up-front what the bounds are, and while loops better for searches - i.e., running until some condition is met.